centralstore/redis_store.go

```diff
@@ -745,29 +737,89 @@ func (t *tracesStore) keepTrace(ctx context.Context, conn redis.Conn, status []*
 	return nil
 }
 
-func (t *tracesStore) getTraceStatuses(ctx context.Context, conn redis.Conn, traceIDs []string) (map[string]*CentralTraceStatus, error) {
+func (t *tracesStore) getTraceStatuses(ctx context.Context, client redis.Client, traceIDs []string) (map[string]*CentralTraceStatus, error) {
 	_, statusSpan := otelutil.StartSpanWith(ctx, t.tracer, "getTraceStatuses", "num_ids", len(traceIDs))
 	defer statusSpan.End()
 
 	if len(traceIDs) == 0 {
 		return nil, nil
 	}
 
-	statuses := make(map[string]*CentralTraceStatus, len(traceIDs))
-	for _, traceID := range traceIDs {
-		status := &centralTraceStatusRedis{}
-		err := conn.GetStructHash(t.traceStatusKey(traceID), status)
-		if err != nil {
-			if errors.Is(err, redis.ErrKeyNotFound) {
-				continue
-			}
-			statusSpan.RecordError(err)
-			return nil, fmt.Errorf("failed to retrieve trace status for trace ID %s with error %s", traceID, err)
+	fanoutChan := make(chan string, 100)
+	faninChan := make(chan *CentralTraceStatus, 100)
+
+	// send all the trace IDs to the fanout channel
+	wgFans := sync.WaitGroup{}
+	wgFans.Add(1)
+	go func() {
+		defer wgFans.Done()
+		defer close(fanoutChan)
+		for i := range traceIDs {
+			fanoutChan <- traceIDs[i]
 		}
+	}()
 
-		statuses[traceID] = normalizeCentralTraceStatusRedis(status)
+	// collect the statuses from the fanin channel
+	statuses := make(map[string]*CentralTraceStatus, len(traceIDs))
+	wgFans.Add(1)
+	go func() {
+		defer wgFans.Done()
+		for status := range faninChan {
+			statuses[status.TraceID] = status
+		}
+	}()
+
+	// determine the number of goroutines to use based on the number of traceIDs
+	numGoroutines := len(traceIDs)/10 + 1
+	if numGoroutines > 10 {
+		numGoroutines = 10
+	}
+
+	// create workers to get the traceIDs and send their statuses to the fanin channel
+	// They will pull from the fanout channel and terminate when it closes
+	wgWorkers := sync.WaitGroup{}
+	for i := 0; i < numGoroutines; i++ {
+		wgWorkers.Add(1)
+		go func(i int) {
+			conn := client.Get()
+			defer conn.Close()
+			queries := 0
+			found := 0
+			_, span := otelutil.StartSpanWith(ctx, t.tracer, "getTraceStatusesWorker", "worker", i)
+			defer span.End()
+			defer wgWorkers.Done()
+			for traceID := range fanoutChan {
+				status := &centralTraceStatusRedis{}
+				err := conn.GetStructHash(t.traceStatusKey(traceID), status)
+				queries++
+				if err != nil {
+					if errors.Is(err, redis.ErrKeyNotFound) {
+						continue
+					}
+					statusSpan.RecordError(err)
+					continue
+				}
+				faninChan <- normalizeCentralTraceStatusRedis(status)
+				found++
+				// just to make sure we give all the goroutines a chance to run
+				time.Sleep(1 * time.Microsecond)
```

Comment:
Wow, does the go scheduler not able to manage this appropriately?

Reply:
Oh! I can remove this now; I was trying to figure out why I only got 3 goroutines running, and then realized it was because the redis connection pool was so limited.